### ISSUE TITLE
PHP API to work with Ext JS Examples

### ISSUE THREAD

**Issue by NilsBe** [mevdschee/php-crud-api#119]
I have chosen your PHP REST API as it seems to be simple and lightweight but mature enough for my first steps with Ext JS. 

Unfortunately it doesn't work. And I would like to focus more on learning Ext JS and UI creating, than REST or SOAP or CRUD. 

I have also found another REST API which works, but it supports only reading. 

Here's a comparison: 

Your API
  {"users":{"columns":["id","email","first","last"],"records":[["0","e","f","l"],["1","e5","f5","l5"],["2","e6","f6","l6"],["3","e7","f7","l7"]]}}

and this is, what Ext JS wants:
  [{"id":"0","email":"e","first":"f","last":"l"},{"id":"1","email":"e5","first":"f5","last":"l5"},{"id":"2","email":"e6","first":"f6","last":"l6"},{"id":"3","email":"e7","first":"f7","last":"l7"}]

Is there a way to change your API to create the output in the right form?

Thanks a lot!

**Comment by mevdschee** [comment]
I think you should add "&transform=1" to the url and set the root to "users" for both the reader and the writer.

**Comment by mevdschee** [comment]
If you send me a code snippet from what you have now I can add it to "examples" and help you adjust it so that it works.

**Comment by NilsBe** [comment]
Dear Maurits, 
I didn't answer to you immediately because I found out, that your minimalistic version of a REST API from here https://www.leaseweb.com/labs/2015/10/creating-a-simple-rest-api-in-php/ is already fully helpful for me. Thanks a lot for that!

But if I would like to use this API, how actually do I set the root to users?

kind regards
Nils

**Comment by mevdschee** [comment]
Well that can be seen here:

http://stackoverflow.com/questions/14841040/sencha-extjs-restful-grid-example-confusion

where it is set to "data".

**Comment by NilsBe** [comment]
Thank you!
It works now.

[rest.zip](https://github.com/mevdschee/php-crud-api/files/563457/rest.zip)

**Comment by mevdschee** [comment]
Great!

**Comment by NilsBe** [comment]
Thanks to your API and your support.
I have seen you also answered several questions on stackexchange. 
Unfortunately this is not as appreciated as I think it should.

**Comment by mevdschee** [comment]
Thank you very much for your efforts!

**Comment by fairport** [comment]
Maurits &/or Nils ..
I am in the same or similar boat as Nils was. I have tried both sets of api.php code.
I have hard wired $transform = true; into the function listCommand .. which has got me closer. But in the proxy reader of ExtJS6 the  API is returning 
a result of  {AssetTypes: Array[5], ClientID: "", dbPostDate: null, id: "AssetTypes-1", AssetTypeID: "}
and the AssetType array is what I really want! 
In the proxy config in ExtJS I tried root: 'AssetTypes' & rootProperty: 'AssetTypes'

So... the data is coming OK, just in the wrong format.
I am a total newbie in this department... so forgive my possible ignorance. 

Thnx in advance
Roger 
P.S. this api code is exactly what I have spent a week searching for!! so I'd be very pleased to get it fully operational

**Comment by mevdschee** [comment]
Hi fairport, I want to help you, but you need to be (much) more specific. You seem to be suffering from the same issue that @NilsBe has, but apparently the same solution does not work for you, right? If you provide reproduction steps then I might be able to help you or if your API is public I can help you debug it. Anyway, I would love to help you, but with your brief problem description I can't.

**Comment by fairport** [comment]
Thanks Maurits
Being a novice at this, I am following the basic requirements for communicating with a rest api in ExtJS6. ExtJS constructs the URL dynamically.
http://docs.sencha.com/extjs/6.2.0/classic/Ext.data.proxy.Rest.html

I am using mySQL
I tweaked the code in Nil's rest.zip thus:
```
protected function listCommand($parameters) {
		extract($parameters);
		$this->startOutput($callback);
		$transform = true;  <<<<<
	        ..... etc.
```
which had the same effect as add ?transform=1 .  to the URL 
localhost/myapp/AssetTypes?transform=1    (I actually get 5 records .. which is expected)
{"AssetTypes":[{"ClientID":"b32538b4-cb7c-11e6-b2a0-7f6a95ada327","AssetTypeID":"8cdeda48-cb7c-11e6-b2a0-7f6a95ada327","AssetType":"Car","dbPostDate":"2016-12-26 22:35:42","dbPostUser":null}]}

What ExtJS wants is [{"ClientID":"b32538b4-cb7c-11e6-b2a0-7f6a95ada327","AssetTypeID":"8cdeda48-cb7c-11e6-b2a0-7f6a95ada327","AssetType":"Car","dbPostDate":"2016-12-26 22:35:42","dbPostUser":null}]

I Hope that helps.
Roger

**Comment by mevdschee** [comment]
I think that you should set the option 'root' to 'AssetTypes' for the 'reader'.

See: http://stackoverflow.com/questions/14841040/sencha-extjs-restful-grid-example-confusion

**Comment by fairport** [comment]
Yes, I _thought_ I'd tried that already & 'rootProperty' .. both made no difference. 
Now I see why !! I needed to set that property in the reader section!!
thus:
```
_proxy:{
        url: host+'/AssetTypes',
        idParam: 'AssetTypeID',
        reader: {
            rootProperty: 'AssetTypes'
        }
    }_
```

Sorry & Thanks for you time Maurits!

Regards
Roger

**Comment by mevdschee** [comment]
I'm glad you worked it out. Enjoy the software!

**Comment by fairport** [comment]
Maurits
FYI & for the benefit of other ExtJS 6 users who would like to use your api... 
I had to override 2 ExtJS classes to get this working.

a) On a PUT or a DELETE a null is received in the responseText with quotes around it. ExtJS mishandles that, so I had to override the 'Ext.data.reader.Reader' read function ...

b)  The dynamic URL builder needs overriding to comply with your requirements .. ExtJS doesn't seem to prepare the URL the way you need it.

Override files attached. in a zip.
[overrides for ExtJS.zip](https://github.com/mevdschee/php-crud-api/files/678534/overrides.for.ExtJS.zip)


Rgds
Roger

**Comment by mevdschee** [comment]
@fairport: Thank you very much! I will look into it. I think "a" actually is a bug. I will investigate. Can you explain when that happens exaclty (do you have an example)?

**Comment by fairport** [comment]
It was happening on a "PUT". I had constructed the URL and on inspection, the response object had the property response.responseText = 'null'.   I suspect if it was null (without quotes) it would have not caused the error. 

As it happens Maurits, my IT guru here has decided to write a dedicated api for this application we are writing... so we have moved on.

**Comment by mevdschee** [comment]
Thank you.

>my IT guru here has decided to write a dedicated api for this application we are writing... so we have moved on.

Was that a case of "Not Invented Here" or ensuring employment or were there specific reasons to choose that route?

**Comment by fairport** [comment]
Hi Maurits

Over the Christmas break I researched the options that we could/should 
use along with ExtJS.
I concluded from my research REST was the way to go as opposed to some 
hack using Ext.Direct or Ajax.

He has not been as closely involved with ExtJS as I have & in his brief 
time googling for a solution he came across a tutorial on Ext.Direct and 
thought that was the way to go.

When I convinced him REST was best & Ext.Direct was built for a 
different purpose (even though it can be used for database CRUD work), 
he accepted my conclusion & then said we could use the REST api he 
developed last year for another project.

So, long story short, he was a) already quite familiar with REST, b) he 
had a system already working & was handling stored procedures and other 
tricks like multiple AND's & OR's in WHERE clauses

We are late on scene in web app development although we have developed 
some radical data sync. systems for PDA's since 2001, & recently smart 
phones since 2011... All our mobile apps to date have run offline & push 
& pull data on user demand.

We're on the way now... & your api was a great help in that. This 
current project is a pilot project for a very big one that comes next... 
so I think to have our own in-house api is better at the end of the day.

Regards
Roger





------ Original Message ------
